### ISSUE TITLE
[Bug] Typescript support for workspaces

### ISSUE THREAD

**Issue by alubbe** [yarnpkg/berry#1483]
**Describe the bug**

I'm not sure if this is a yarn 2 bug or a CRA bug, but I cannot use typescript with CRA when using workspaces (it works just fine without workspaces).

**To Reproduce**
See https://github.com/alubbe/yarn2_workspace_typescript

Here's the version, without workspaces, that works just fine: https://github.com/alubbe/yarn2_workspace_typescript/tree/5a902ae8e3b39aa33e9c84ed0000b88fdc7bb9c5
The repo was created by these commands:
```
npx create-react-app pnp --use-pnp --typescript
yarn set version berry
yarn install
// ... add `connected-react-router`
```

Then, I simply move the files into a workspace (https://github.com/alubbe/yarn2_workspace_typescript/commit/7839493116b4df41a7dfd328b9819116d20d1cff), and suddenly I get this error: 
```
TypeScript error in undefined(undefined,undefined):
Cannot find type definition file for 'node'.
```
So I fix `ts-pnp` by unplugging (see https://github.com/arcanis/pnp-webpack-plugin/issues/14#issuecomment-534132701 and https://github.com/alubbe/yarn2_workspace_typescript/commit/4c062a19b5515853699ccda95d671b3e990bb6e7), but I cannot get past this error:
```
TypeScript error in /packages/rofl/src/index.tsx(10,5):
JSX element class does not support attributes because it does not have a 'props' property.  TS2607

     8 | ReactDOM.render(
     9 |   <React.StrictMode>
  > 10 |     <ConnectedRouter history={{}}>
       |     ^
    11 |       <App />
    12 |     </ConnectedRouter>
    13 |   </React.StrictMode>,
```
So, it looks like TypeScript can find types from @types just fine, but not from a package itself (`connected-react-router` has its own types file). Am I doing something wrong or is this a bug?

**Comment by backslash47** [comment]
Hello, I ended up solving similar problem with 
`JSX element class does not support attributes because it does not have a 'props' property` while using react-intl and react-draggable with yarn2 pnp and workspaces by adding:

```yaml
packageExtensions:
  react-intl@*:
    peerDependenciesMeta:
      "@types/react":
        optional: true
  react-draggable@*:
    peerDependenciesMeta:
      "@types/react":
        optional: true
```

to yarnrc.yml

Hope it helps.

**Comment by alubbe** [comment]
As I'm struggling using sherlock (see https://github.com/yarnpkg/berry/issues/1433), is there anything else I can add here? Again, it should be able to reproduce the issue here by cloning https://github.com/alubbe/yarn2_workspace_typescript and running `yarn install && cd packages/rofl && yarn start`

**Comment by alubbe** [comment]
I hate to bump my own ticket, but this issue is causing major headaches for us - is there anything we can do to help investigate this further? Is this even the right place to report the issue? Why does TS find all the JS files just fine, but struggle to find a dependency's own type declaration file? Any pointers would be extremely helpful!

**Comment by merceyz** [comment]
Sorry about not getting to this sooner, I looked into it and the issues here are:
- The version of `ts-pnp` is outdated, you fixed this by unplugging and applying the fix manually, however you can also set this in your `package.json`
```
"resolutions": {
  "ts-pnp": "^1.2.0"
}
```
- `connected-react-router` is missing some dependencies (as hinted to by @backslash47 in https://github.com/yarnpkg/berry/issues/1483#issuecomment-649566062)
You can fix these by adding this to your `.yarnrc.yml` file
```
packageExtensions:
  "connected-react-router@*":
    dependencies:
      "@types/react": "*"
      "@types/redux": "*"
      "@types/react-redux": "*"
      "@types/react-router": "*"
      "@types/history": "*"  
```
The reason it works without workspaces is that you declared these dependencies yourself at the root and because of the fallback (https://yarnpkg.com/configuration/yarnrc#pnpFallbackMode) it was allowed to use those

**Comment by arcanis** [comment]
> `connected-react-router` is missing some dependencies (as hinted to by @backslash47 in [#1483 (comment)](https://github.com/yarnpkg/berry/issues/1483#issuecomment-649566062))

Of note, I'm working on a fix that will make peer dependencies imply the associated @types optional peer dependency. In this context, you'll have to add the @types packages to your own dependencies without having to fix the extensions through the configuration.

**Comment by alubbe** [comment]
Thank you all - you've helped me resolve the issue! Two quick notes for anyone else finding this thread in the future:
1. While `@types/redux` still exists and using it fixed the error, it is deprecated and its types are no longer compatible with `redux`. I got a complicated type error later on because `redux` now ships its own types file - so I just replaced `@types/redux` with `redux` as a dependency of `connected-react-router`
2. After everything was building, I noticed that (at least for our code base) `connected-react-router` only seemed to need the extra `@types/react` dependency. So for us, this was the final fix:
```yaml
  connected-react-router@^6.5.2:
    dependencies:
      "@types/react": ^16.9.2
```
